lib/dartdoc.dart

```diff
@@ -167,21 +185,170 @@ class DartDoc {
 
     for (var generator in generators) {
       await generator.generate(package, outputDir);
+      generator.writtenFiles
+          .forEach((p) => writtenFiles.add(path.normalize(p)));
     }
 
+    await verifyLinks(package, outputDir.path);
+
     double seconds = _stopwatch.elapsedMilliseconds / 1000.0;
     print(
         "\nDocumented ${package.libraries.length} librar${package.libraries.length == 1 ? 'y' : 'ies'} "
         "in ${seconds.toStringAsFixed(1)} seconds.");
+    print(
+        "Finished with:  ${package.packageWarningCounter.warningCount} warnings, ${package.packageWarningCounter.errorCount} errors");
 
     if (package.libraries.isEmpty) {
-      stderr.write(
-          "\ndartdoc could not find any libraries to document. Run `pub get` and try again.");
+      throw new DartDocFailure(
+          "dartdoc could not find any libraries to document. Run `pub get` and try again.");
+    }
+
+    if (package.packageWarningCounter.errorCount > 0) {
+      throw new DartDocFailure("dartdoc encountered errors while processing");
     }
 
     return new DartDocResults(packageMeta, package, outputDir);
   }
 
+  void _warn(Package package, PackageWarning kind, String p, String origin,
+      {String source}) {
+    // Ordinarily this would go in [Package.warn], but we don't actually know what
+    // ModelElement to warn on yet.
+    Locatable referenceElement;
+    Set<Locatable> referenceElements;
+
+    // Make all paths relative to origin.
+    if (path.isWithin(origin, p)) {
+      p = path.relative(p, from: origin);
+    }
+    if (source != null) {
+      if (path.isWithin(origin, source)) {
+        source = path.relative(source, from: origin);
+      }
+      // Source paths are always relative.
+      referenceElements = package.allHrefs[source];
+    } else {
+      referenceElements = package.allHrefs[p];
+    }
+    if (referenceElements != null) {
+      if (referenceElements.any((e) => e.isCanonical)) {
+        referenceElement = referenceElements.firstWhere((e) => e.isCanonical);
+      } else {
+        // If we don't have a canonical element, just pick one.
+        referenceElement =
+            referenceElements.isEmpty ? null : referenceElements.first;
+      }
+    }
+    if (referenceElement == null && source == 'index.html')
+      referenceElement = package;
+    package.warn(referenceElement, kind, p);
+  }
+
+  Future _doOrphanCheck(
+      Package package, String origin, Set<String> visited) async {
+    String normalOrigin = path.normalize(origin);
+    String staticAssets = path.joinAll([normalOrigin, 'static-assets', '']);
+    String indexJson = path.joinAll([normalOrigin, 'index.json']);
+    bool foundIndex = false;
+    await for (FileSystemEntity f
```

Comment:
And - not necessarily for this PR - it would be interesting to add the lint that checks that you're only awaiting Futures. It _should_ then fire when you make `_doOrphanCheck` sync, on the await in verifyLinks.